addon/lib/HiddenFrame.jsm

```diff
@@ -0,0 +1,116 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+"use strict";
+
+this.EXPORTED_SYMBOLS = ["HiddenFrame"];
+
+const { classes: Cc, interfaces: Ci, utils: Cu } = Components;
+
+Cu.import("resource://gre/modules/PromiseUtils.jsm");
+Cu.import("resource://gre/modules/Services.jsm");
+Cu.import("resource://gre/modules/XPCOMUtils.jsm");
+
+const XUL_PAGE = "chrome://global/content/win.xul";
+
+const gAllHiddenFrames = new WeakSet();
+
+let cleanupRegistered = false;
+function ensureCleanupRegistered() {
+  if (!cleanupRegistered) {
+    cleanupRegistered = true;
+    Services.obs.addObserver(function() {
+      for (let hiddenFrame of ChromeUtils.nondeterministicGetWeakSetKeys(gAllHiddenFrames)) {
+        hiddenFrame.destroy();
+      }
+    }, "xpcom-shutdown");
+  }
+}
+
+
+/**
+ * An hidden frame object. It takes care of creating a windowless browser and
+ * passing the window containing a blank XUL <window> back.
+ */
+function HiddenFrame() {
```

Comment:
Is there a reason not to use an es6 class here? Other parts of the code do, and mixing them both seems confusing.

Reply:
This file comes from the firefox source tree [here](https://searchfox.org/mozilla-central/source/toolkit/modules/HiddenFrame.jsm). However, I cannot use it directly since it is not exported anymore outside of tests. For this reason, I decided to ship it with the addon. I wouldn't want to change it though, as leaving it identical preserves the assumptions made by the test coverage in-tree (I guess) :D

Reply:
That's a great reason to leave the code unmodified. Carry on 😄